protocol/0042-LIQF-setting_fees_and_rewarding_lps.md

```diff
@@ -193,53 +201,104 @@ Then on every Vega time change, after `fractional instantaneous liquidity score`
+The liquidity fees are transferred from the market LP fee account into the LP-per-market fee account, pro-rata depending on the `LP i equity-like share` multiplied by `LP i liquidity score` scaled back to `1` across all LPs at a given time.
 
-The liquidity fees are distributed pro-rata depending on the `LP i equity-like share` multiplied by `LP i liquidity score` scaled back to `1` across all LPs at a given time.
+The LP parties don't control the LP-per-market fee account; the fees from there are then tranferred to the LPs' general account at the end epoch as described below.
 
-#### Example
 
-We have `4` LPs with equity-like share shares:
-share as below
 
-```math
-LP 1 els = 0.65
-LP 2 els = 0.25
-LP 3 els = 0.1
-```
+### Calculating SLA performance
 
-Trade happened, and the trade value for fee purposes multiplied by the liquidity fee factor is `103.5 ETH`. The following amounts be collected immediately into the LP fee accounts for the market:
+#### Measuring time spent meeting their commitment in a single epoch
 
-```math
-0.65 x 103.5 = 67.275 ETH to LP 1's LP account
-0.25 x 103.5 = 25.875 ETH to LP 2's LP account
-0.10 x 103.5 = 10.350 ETH to LP 3's LP account
-```
+During the epoch, the amount of time in nanoseconds (of Vega time) that each LP spends meeting the SLA is recorded. This can be done by maintaning a counter `s_i` as shown below:
+
+* At the start of a new epoch, `s_i = 0`
+
+    * If the LP is meeting their commitment, store the Vega time of the start of the epoch as the time the LP began meeting their commitment, otherwise store `nothing`.
+
+* At the end of each block:
+
+    * If LP has started meeting their [committed volume of notional](./0044-LIME-lp_mechanics.md) (section "Calculating liquidity from commitment") after previously not doing so (i.e. `nothing` is stored as the time the LP began meeting their commitment):
+
+        * Store the current Vega time as the time the LP began meeting their commitment
 
-Then LP 4 made a delayed LP commitment, and updated share as below:
+    * If LP has stopped meeting their committed volume of notional after previously doing so:
 
-```math
-LP 1 els = 0.43
-LP 2 els = 0.17
-LP 3 els = 0.07
-LP 4 els = 0.33
+        * Add the difference in nanoseconds between the current Vega time and the time the LP began meeting their commitment (stored in the step above) to `s_i`
+        
+        * Store `nothing` as the time the LP began meeting their commitment, to signify the LP not meeting their commitment
+
+* At the end of the epoch, calculate the actual observed epoch length `observed_epoch_length` = the difference in nanoseconds between the Vega time at the start of the epoch and the Vega time at the end of the epoch.
+
+Note that we only need to evaluate each LP's status at the end of each block, as no _Vega time_ passes between transactions in a block.
```

Comment:
Are we safeguarding against an LP trying to place the transactions so that they cancel their orders around the first transaction in the block an reinstate them towards the end of the block (so as to appear as having been actively providing liquidity throughout the block)? 
This may be covered, but I can't convince myself, so might be worth spelling it out.

Reply:
I don't think we are and I can't quite work out how big a problem this is, but I think probably big enough to do something about…

I think the LP's success rate in timing this is key to how effective an attack it might be. Each block where you "miss" re-upping your liquidity at the end will take you off book until the next block, so you can have `1-SLA%` chance of missing before you miss the SLA anyway, which may not be realistic. Though you would potentially manage to reduce your penalty even in this case.

That's assuming you're dropping fully to zero. There's also a chronic underprovision attack which I think is probably simpler and made easier by `refresh policy = BLOCK_END` glassbergs, as these help you to always be meeting the provision at the end of the block without doing so during it.

We could instead evaluate LPs' SLA achievement after each transaction (1=SLA met, 0=not met) and use the average of these for the block multiplied by the block length in nanos to decide how much additional time the LP spent meeting their commitment "during" the block.

This would seem strange in a way given how time evolves vs. transactions, but might be fairer. The strange thing is that the liquidity that's "visible" (i.e. on the book between blocks) exists for all the "Vega time" (and is what people see) but a tiny fraction of the considered provision. It would also make `refresh policy = BLOCK_END` pretty useless, so if we do this I'd propose removing that (in fact that may be an _enabler_ of under-provision in the way you describe). 

This "solution" would create an incentive to supply liqudity then make many transactions that don't trade with it or only trade a little, though. That could perhaps be solved by excluding transactions that don't trade on the market, and weighting the average described above by the volume traded by each transaction.

Either solution increases the amount of work done per LP per blockto measure SLA performance from from O(1) to O(n) (n of [trading] transactions), and increases the complexity of the calculation a bit too.

How else might we solve this?

Reply:
Coule we use the minimum liquidity within block? We'd then need to monitor all cancels and amends within the block and exclude glassberg fills from updating the minimum.

Reply:
So really just calculate it at the beginning of the block and monitor cancels, amends and fills (of non-glassberg orders) within the block to update the minimum.

Reply:
Or we just don't care about the fills at all, but not sure if it doesn't open a possibility of wash trading back and forth between two LPs.

Reply:
@davidsiska-vega came up with a much better and simpler solution:

We randomise the point in the block at which we determine whether an LP met the SLA for that block.

So for a block containing _N_ transactions (not _N_ orders, a batch is 1 txn), we pick `random_integer_between(0, N)` and asses the LP's liquidity after that transaction. We could seed the RNG with something like the hash of all transaction hashes in the block or the previous state/block hash.

Glassbergs with `BLOCK_END` would still refresh at block end, so to be certain they are providing some amount of liquidity every time you would need to use that as `minimum peak size` rather than `initial peak size`. I think this is ok.

Reply:
I have added the above to the PR.

Reply:
> Coule we use the minimum liquidity within block? We'd then need to monitor all cancels and amends within the block and exclude glassberg fills from updating the minimum.

We _could_ use their minimum. I think that is somewhat harsh compared to what MMs will be used to. IIRC CEXs sample the data (less often than per block, even) cc @MM0819 is that correct?

Reply:
They are roughly doing this:

Randomly sample your orders on the book within 0.3% (for example) twice per minute. Then take the average of all of these samples over the day, and require the average to be > $100k (again, for example). You retain the additional MM benefits if you satisfy this requirement on more than 90% of the days within the month. In the case of CEXs, this will be negative maker fees.

This is nice because it allows you to quote e.g. $1M of depth during times of low volatility (weekends etc), and quote $0 during times of predictably high volatility (e.g. non-farm payrolls, CPI, FOMC etc).

Reply:
I should add that the penalty is quite harsh for not meeting these requirements. If you do not satisfy these requirements in two consecutive months you are kicked out of the market making program completely... :smile:

Reply:
So, if we propose to have less strict penalties, we can surely have stricter requirements.

Reply:
The actual strictness (amounts, depth, distance from best price) will be configurable by market and in any case the rewards will be competitive.

So the SLA to not get a penalty against you will be pretty permissive, but if others are consistently more competitive than you you’ll stay to get less reward pretty quickly (like a more advanced version of your “if they commit $10 I’ll commit $1000 and take all their fees” but measuring their actual performance on the SLA KPIs).